src/NServiceBus.Core/MessageHandlerContext.cs

```diff
@@ -0,0 +1,69 @@
+namespace NServiceBus
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Threading.Tasks;
+    using NServiceBus.Transports;
+    using NServiceBus.Unicast;
+
+    class MessageHandlerContext : IMessageHandlerContext
+    {
+        ContextualBus bus;
+
+        public MessageHandlerContext(ContextualBus bus)
+        {
+            this.bus = bus;
+        }
+
+        public string MessageId => bus.MessageBeingProcessed.MessageId;
+
+        public string ReplyToAddress => bus.MessageBeingProcessed.GetReplyToAddress();
+
+        public IReadOnlyDictionary<string, string> MessageHeaders => bus.MessageBeingProcessed.Headers;
```

Comment:
@timbussmann  but this doesnt new up a collection each time? it is just returning the message headers